llmfoundry/utils/config_utils.py

```diff
@@ -25,7 +28,258 @@
+        ret = om.to_container(cfg, resolve=True)
+        assert isinstance(ret, list)
+        return ret  # type: ignore (see above)
+    else:
+        return cfg  # type: ignore (dicts and lists are already in the correct format)
+
+
+T = TypeVar('T')
+
+
+def make_dataclass_and_log_config(
+        cfg: DictConfig,
+        dataclass_constructor: Callable[..., T],
+        dataclass_fields: Set[str],
+        transforms: Optional[List[Callable[[Dict[str, Any]],
+                                           Dict[str, Any]]]] = None,
+        icl_tasks_required: bool = False) -> Tuple[Dict[str, Any], T]:
+    """Converts a DictConfig to a dataclass and creates a logged config."""
+    # Resolve all interpolation variables as early as possible
+    unstructured_config = om.to_container(cfg, resolve=True)
+    assert isinstance(unstructured_config, dict)
+    assert all(isinstance(k, str) for k in unstructured_config.keys())
+    unstructured_config = {str(k): v for k, v in unstructured_config.items()}
+
+    # Flatten union types before creating structured config:
+    if 'eval_gauntlet' in unstructured_config:
+        forbid_config_key(unstructured_config, 'eval_gauntlet_str')
+        if isinstance(unstructured_config['eval_gauntlet'], str):
+            unstructured_config['eval_gauntlet_str'] = unstructured_config.pop(
+                'eval_gauntlet')
+    if (loader := unstructured_config.get('eval_loader', None)) is not None:
+        forbid_config_key(unstructured_config, 'eval_loaders')
+        if isinstance(loader, list):
+            unstructured_config['eval_loaders'] = unstructured_config.pop(
+                'eval_loader')
+    if 'icl_tasks' in unstructured_config:
+        forbid_config_key(unstructured_config, 'icl_tasks_str')
+        if isinstance(unstructured_config['icl_tasks'], str):
+            unstructured_config['icl_tasks_str'] = unstructured_config.pop(
+                'icl_tasks')
+    else:
+        if icl_tasks_required:
+            raise MissingMandatoryValue(
+                'icl_tasks must be specified in the config')
+
+    # Create copy of config for logging
+    logged_cfg: Dict[str, Any] = copy.deepcopy(unstructured_config)
+
+    # apply transforms to the unstructured config before constructing dataclass
+    for transform in transforms or []:
+        unstructured_config = transform(unstructured_config)
+
+    logged_cfg.update(unstructured_config, merge=True)
+
+    arg_config_keys = set(unstructured_config.keys())
+    extraneous_keys = set.difference(arg_config_keys, dataclass_fields)
+
+    if 'variables' not in unstructured_config:
+        unstructured_config['variables'] = {}
+
+    for key in extraneous_keys:
+        warnings.warn(
+            f'Unused parameter {key} found in cfg. Please check your yaml to ensure this parameter is necessary. Interpreting {key} as a variable for logging purposes. Top-level variables are deprecated and will not be supported in future releases.',
```

Comment:
warning should also say that you can put variables in the variables section to indicate that they are intentional variables